src/core/net/dnssd_server.cpp

```diff
@@ -921,6 +949,154 @@ bool Server::Response::QueryNameMatchesService(const Srp::Server::Service &aServ
 
 #endif // OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
 
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE || OPENTHREAD_CONFIG_DNSSD_DISCOVERY_PROXY_ENABLE
+
+void Server::ConstructSoaServerName(void)
+{
+    if (mSoaServerName[0] == kNullChar)
+    {
+        StringWriter writer(mSoaServerName, sizeof(mSoaServerName));
+
+        writer.Append("otDNS%s", Get<Mac::Mac>().GetExtAddress().ToString().AsCString());
+    }
+}
+
+Error Server::Response::ResolveSoaOrNsQuery(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(QueryNameIsForDomain(kDefaultDomainName));
+
+    mSection = kAnswerSection;
+
+    // Handle AAAA (or ANY) query for the server host name used for
+    // SOA/NS answer.
+
+    if (mQuestions.IsFor(kRrTypeAaaa) || mQuestions.IsFor(kRrTypeAny))
+    {
+        Name::Buffer serverFullName;
+
+        ConstructFullName(Get<Server>().mSoaServerName, serverFullName);
```

Comment:
Just as a consideration: since this part is not a SOA or NS query, it would also sound logical to attempt this AAAA (or ANY) record resolution as a last resort, after the "SRP server" resolving part did not produce any results. Then we would also keep the function `ResolveSoaOrNsQuery` more strictly to SOA or NS queries (including ANY query on the zone name).  But overall, it looks ok to me like this as well.

Reply:
Order should not matter. Prefer to keep all related stuff for this feature together within one method.